File: _stats_build/statistics/wr_dominance.rb
# NOTE: Dominance（屠榜）= 选手在某项目全历史成绩排行榜上完全霸占前 N 席
#
# 概念说明:
#   将某项目所有历史 results（每轮的 best 或 average）按成绩值升序排列，
#   如果排行榜前 N 名全部属于同一个人，则该人的 dominance = N。
#   并列成绩排除——只算严格优于他人最佳的部分。
#
# 算法:
#   1. 找到全局成绩最好的选手 P（best_by_person 的最小值）
#   2. 找到非 P 选手中的最佳成绩 others_best（second_best）
#   3. dominance = P 的成绩中严格 < others_best 的数量
#
# WR 历史:
#   按比赛日期正序处理所有 results，每场比赛结束后重新计算 dominance。
#   只记录 dominance 刷新历史最高的时刻，最终倒序输出。
#
# 性能优化:
#   - 每人的成绩用有序数组存储（二分插入 O(log n)）
#   - dominance 计数用 binary search（O(log n)），避免遍历全量
#   - 维护 best_by_person hash 快速找 top/second（O(p)，p=选手数）
#
# 设计决策:
#   - 提供双视图：当前排名 + WR 历史
#   - Single 和 Average/Mean 均覆盖所有项目（含退役项目如脚拧、八板等）
require_relative "../core/statistic"
require_relative "../core/events"
require_relative "../core/solve_time"
require_relative "../core/database"
require_relative "../core/tab_ui"
require_relative "../core/metric_selector"

class WrDominance < Statistic
  include TabUi
  include MetricSelector

  # NOTE: WR 历史表头
  HISTORY_HEADER = {
    "Count" => :right, "Improvement" => :right, "Days" => :right,
    "Person" => :left, "Start Date" => :left, "Start Comp" => :left,
    "Date" => :left, "Competition" => :left
  }.freeze

  # NOTE: 当前排名表头
  RANKING_HEADER = {
    "#" => :right, "Person" => :left, "Count" => :right,
    "Date" => :left, "Competition" => :left
  }.freeze

  def initialize
    @title = "Dominance"
    @note = "A competitor completely dominates top N on the leaderboard of results. Tied results are excluded."
  end

  def query
    ""
  end

  def data
    @data ||= compute_all
  end

  def markdown
    md = top
    md += metric_selector_styles
    md += dominance_metric_buttons

    # NOTE: Tab 样式（全局只输出一次）
    md += tab_styles

    # NOTE: Single 面板——内含 Tab 双视图（当前排名 + WR 历史）
    md += "<div class=\"metric-panel active\" id=\"metric-single\">\n"
    md += tab_buttons(
      "Current Ranking", "当前排名", "single-ranking",
      "WR History", "WR 历史", "single-history"
    )
    md += grouped_panel("single-ranking", true, data[:single][:ranking].to_h, RANKING_HEADER)
    md += grouped_panel("single-history", false, data[:single][:history].to_h, HISTORY_HEADER)
    md += "</div>\n"

    # NOTE: Average 面板
    md += "<div class=\"metric-panel\" id=\"metric-average\">\n"
    md += tab_buttons(
      "Current Ranking", "当前排名", "average-ranking",
      "WR History", "WR 历史", "average-history"
    )
    md += grouped_panel("average-ranking", true, data[:average][:ranking].to_h, RANKING_HEADER)
    md += grouped_panel("average-history", false, data[:average][:history].to_h, HISTORY_HEADER)
    md += "</div>\n"

    md += metric_selector_script
    md += tab_script
    md
  end

  private

  # NOTE: 指标按钮（Single / Average）
  def dominance_metric_buttons
    html = "<div class=\"metric-selector\">\n"
    [{"Single" => "single"}, {"Average" => "average"}].each_with_index do |m, i|
      label, id = m.first
      active = i == 0 ? " active" : ""
      html += "  <button class=\"metric-btn#{active}\" onclick=\"switchMetric('#{id}')\" "
      html += "data-i18n-en=\"#{label}\">#{label}</button>\n"
    end
    html += "</div>\n"
    html
  end



  def compute_all
    result = {
      single:  { ranking: [], history: [] },
      average: { ranking: [], history: [] }
    }

    puts "  Dominance: querying all results..."
    all_rows = fetch_all_results

    # Single: 所有项目
    Events::ALL.each do |event_id, event_name|
      rows = all_rows.select { |r| r["event_id"] == event_id && r["best"] > 0 }
      next if rows.empty?

      dom = compute_dominance(rows, "best")
      result[:single][:history] << [event_name, dom[:history]] unless dom[:history].empty?
      result[:single][:ranking] << [event_name, dom[:ranking]] unless dom[:ranking].empty?
    end

    # Average: 所有项目
    Events::ALL.each do |event_id, event_name|
      rows = all_rows.select { |r| r["event_id"] == event_id && r["average"] > 0 }
      next if rows.empty?

      dom = compute_dominance(rows, "average")
      result[:average][:history] << [event_name, dom[:history]] unless dom[:history].empty?
      result[:average][:ranking] << [event_name, dom[:ranking]] unless dom[:ranking].empty?
    end

    result
  end

  # NOTE: 一次查询同时取 best 和 average，由调用方按需使用对应列
  def fetch_all_results
    sql = <<-SQL
      SELECT
        result.event_id,
        result.person_id,
        result.best,
        result.average,
        CONCAT('[', person.name, '](https://www.worldcubeassociation.org/persons/', person.wca_id, ')') person_link,
        CONCAT('[', competition.cell_name, '](https://www.worldcubeassociation.org/competitions/', competition.id, ')') comp_link,
        competition.start_date
      FROM results result
      JOIN persons person ON person.wca_id = result.person_id AND person.sub_id = 1
      JOIN competitions competition ON competition.id = result.competition_id
      WHERE result.best > 0 OR result.average > 0
      ORDER BY result.event_id, competition.start_date
    SQL
    Database.client.query(sql).to_a
  end

  # NOTE: 统一计算 dominance 的 WR 历史和当前排名（共享 pv/pb 数据结构，单次遍历）
  # value_col: "best" 或 "average"，指定从哪个列读取成绩值
  # 返回 { history: [[cols...], ...], ranking: [[cols...], ...] }
  def compute_dominance(rows, value_col)
    # per-person 排序值列表（用于 binary search 计数）
    pv = Hash.new { |h, k| h[k] = [] }  # person_id => sorted [values]
    # 每人最佳（最小）值
    pb = {}  # person_id => best value
    # 每人最新的 link/date 信息（WR 历史和当前排名共用）
    pi = {}  # person_id => { person_link, comp_link, date }

    max_dom = 0
    wr_records = []
    # NOTE: 追踪每位选手首次 dominance 的比赛信息
    first_dom = {}  # person_id => { comp_link:, date: }

    # --- WR 历史追踪：按日期分组逐步构建 pv/pb ---
    rows.group_by { |r| r["start_date"] }.each do |date, comp_rows|
      comp_rows.each do |r|
        pid = r["person_id"]
        val = r[value_col]

        # 二分插入，保持有序
        idx = pv[pid].bsearch_index { |v| v >= val } || pv[pid].size
        pv[pid].insert(idx, val)

        # 更新 best
        pb[pid] = val if !pb[pid] || val < pb[pid]

        pi[pid] = {
          person_link: r["person_link"],
          comp_link: r["comp_link"],
          date: r["start_date"]
        }
      end

      # 找 top person（best 最小的人）和 second_best（非 top 的最小 best）
      top_pid = nil
      top_best = Float::INFINITY
      second_best = Float::INFINITY

      pb.each do |pid, best|
        if best < top_best
          second_best = top_best
          top_pid = pid
          top_best = best
        elsif best < second_best
          second_best = best
        end
      end

      # 如果只有一个人，跳过
      next if second_best == Float::INFINITY

      # 用 binary search 统计 top_person 有多少值 < second_best
      arr = pv[top_pid]
      cnt = arr.bsearch_index { |v| v >= second_best } || arr.size

      if cnt > max_dom && cnt > 0
        max_dom = cnt
        first_dom[top_pid] ||= { comp_link: pi[top_pid][:comp_link], date: date }
        wr_records << {
          count: cnt,
          person_id: top_pid,
          person_link: pi[top_pid][:person_link],
          comp_link: pi[top_pid][:comp_link],
          date: date
        }
      end
    end

    # --- WR 历史：计算 improvement 和 days，然后倒序 ---
    history = wr_records.each_with_index.map do |r, i|
      if i > 0
        prev = wr_records[i - 1]
        improvement = "+#{r[:count] - prev[:count]}"
      else
        improvement = ""
      end

      if i < wr_records.size - 1
        next_r = wr_records[i + 1]
        days = (next_r[:date] - r[:date]).to_i.to_s
      else
        days = (Date.today - r[:date]).to_i.to_s
      end
      first = first_dom[r[:person_id]]
      [r[:count], improvement, days,
       r[:person_link],
       first[:date].strftime("%Y-%m-%d"), first[:comp_link],
       r[:date].strftime("%Y-%m-%d"), r[:comp_link]]
    end.reverse

    # --- 当前排名：复用遍历结束后的 pv/pb 最终状态 ---
    ranking = compute_ranking_from_state(pv, pb, pi)

    { history: history, ranking: ranking }
  end

  # NOTE: 从 pv/pb/pi 最终状态计算当前 dominance Top 10
  def compute_ranking_from_state(pv, pb, pi)
    # 预计算全局 top1 和 second_best
    top_pid = nil
    top_best = Float::INFINITY
    second_best = Float::INFINITY

    pb.each do |pid, best|
      if best < top_best
        second_best = top_best
        top_pid = pid
        top_best = best
      elsif best < second_best
        second_best = best
      end
    end

    return [] if second_best == Float::INFINITY

    # 对每个 person 计算 dominance
    # others_best: 如果 P 是全局最佳选手则为 second_best，否则为 top_best
    ranking = []
    pv.each do |pid, values|
      others_best = (pid == top_pid) ? second_best : top_best

      cnt = values.bsearch_index { |v| v >= others_best } || values.size
      next if cnt <= 0

      info = pi[pid]
      date_str = info[:date].respond_to?(:strftime) ? info[:date].strftime("%Y-%m-%d") : info[:date].to_s
      ranking << [cnt, pid, info[:person_link], date_str, info[:comp_link]]
    end

    # 按 dominance 降序排序，取 Top 10
    ranking.sort_by! { |r| -r[0] }
    ranking.first(10).each_with_index.map do |r, i|
      [i + 1, r[2], r[0], r[3], r[4]]
    end
  end
end

File: trainer/packages/stats-build/src/statistics/wr_dominance.ts
// NOTE: Dominance（屠榜）——选手在某项目全历史成绩排行榜上完全霸占前 N 席
// 与 Ruby _stats_build/statistics/wr_dominance.rb 1:1 对应
// 算法：
//   1. 找到全局成绩最好的选手 P
//   2. 找到非 P 选手中的最佳成绩 others_best
//   3. dominance = P 的成绩中严格 < others_best 的数量（二分搜索）
// 双维度（Single + Average）× 双视图（Ranking + History）
//
// ⚠️ 内存优化（v2）——SQL 不返回 CONCAT 链接字符串
//   原版每行 RowDataPacket 含 person_link (~80B) + comp_link (~100B) = ~180B 额外
//   333 single ~500万行 × 180B = ~900MB 纯字符串。改为仅返 person_id + comp_id，
//   最终结果行（几十条）才 resolveLinks 查出链接。
import { formatDate } from '../core/format_date.js';
import { Statistic } from '../core/statistic.js';
import { EVENTS_ENTRIES, headerZh, eventZh } from '../core/events.js';
import { query as dbQuery } from '../core/database.js';
import type { StatJson, MetricPanel, TableHeader } from '../core/statistic.js';
import type { RowDataPacket } from 'mysql2';

// NOTE: 历史表头
const HISTORY_HEADER: TableHeader = {
  'Count': 'right', 'Improvement': 'right', 'Days': 'right',
  'Person': 'left', 'Start Date': 'left', 'Start Comp': 'left',
  'Date': 'left', 'Competition': 'left',
};

// NOTE: 排名表头
const RANKING_HEADER: TableHeader = {
  '#': 'right', 'Person': 'left', 'Count': 'right',
  'Date': 'left', 'Competition': 'left',
};

// NOTE: computeDominance 的中间结果类型（ID only，无链接字符串）
type DomHistoryEntry = {
  count: number; improvement: string; days: string;
  personId: string; firstCompId: string; firstDate: string; compId: string; date: string;
};
type DomRankingEntry = {
  rank: number; personId: string; count: number; compId: string; date: string;
};

export class WrDominance extends Statistic {
  constructor() {
    super();
    this.title = 'Dominance';
    this.titleZh = '屠榜';
    this.note = 'A competitor completely dominates top N on the leaderboard of results. Tied results are excluded.';
    this.noteZh = '选手在某项目全历史成绩排行榜上完全霸占前 N 席。并列成绩排除。';
  }

  query(): string { return ''; }

  // NOTE: 覆写 toJson——多维度面板输出
  async toJson(): Promise<StatJson> {
    const data = await this.computeAll();

    const buildHeader = (header: TableHeader) =>
      Object.entries(header).map(([label, align]) => ({
        key: label.toLowerCase().replace(/\s+/g, '_'),
        label, labelZh: headerZh(label), align,
      }));

    const buildSections = (grouped: [string, unknown[][]][]) =>
      grouped.filter(([, rows]) => rows.length > 0)
        .map(([title, rows]) => ({ title, titleZh: eventZh(title), rows }));

    const buildMetricPanel = (id: string, labelEn: string, labelZh: string,
      metricData: { ranking: [string, unknown[][]][]; history: [string, unknown[][]][] }): MetricPanel => ({
      id, labelEn, labelZh,
      panels: [
        {
          id: 'ranking', labelEn: 'Ranking', labelZh: '排名',
          header: buildHeader(RANKING_HEADER),
          sections: buildSections(metricData.ranking),
        },
        {
          id: 'history', labelEn: 'History', labelZh: '历史',
          header: buildHeader(HISTORY_HEADER),
          sections: buildSections(metricData.history),
        },
      ],
    });

    const metricPanels: MetricPanel[] = [
      buildMetricPanel('single', 'Single', '单次', data.single),
      buildMetricPanel('average', 'Average', '平均', data.average),
    ];

    return {
      id: this.id,
      title: this.title,
      titleZh: this.titleZh || this.title,
      ...(this.note ? { note: this.note } : {}),
      ...(this.noteZh ? { noteZh: this.noteZh } : {}),
      header: [],
      metricPanels,
    };
  }

  private async computeAll() {
    // NOTE: 中间结果用 ID-only 类型，最终 resolveLinks 后转为 unknown[][]
    const singleHistRaw: [string, DomHistoryEntry[]][] = [];
    const singleRankRaw: [string, DomRankingEntry[]][] = [];
    const avgHistRaw: [string, DomHistoryEntry[]][] = [];
    const avgRankRaw: [string, DomRankingEntry[]][] = [];

    // NOTE: 收集所有出现在结果中的 person/comp ID
    const allPersonIds = new Set<string>();
    const allCompIds = new Set<string>();

    const collectIds = (hist: DomHistoryEntry[], rank: DomRankingEntry[]) => {
      for (const h of hist) {
        allPersonIds.add(h.personId);
        allCompIds.add(h.compId);
        allCompIds.add(h.firstCompId);
      }
      for (const r of rank) {
        allPersonIds.add(r.personId);
        allCompIds.add(r.compId);
      }
    };

    // NOTE: Single 基于所有 individual attempts
    for (const [eventId, eventName] of EVENTS_ENTRIES) {
      const t = Date.now();
      process.stdout.write(`  Dominance single: ${eventId}...`);
      let rows = await this.fetchSingleAttemptsFor(eventId);
      if (rows.length === 0) { console.log(' skip'); continue; }

      const dom = this.computeDominance(rows, 'value');
      rows = null as unknown as RowDataPacket[];
      if (typeof globalThis.gc === 'function') globalThis.gc();

      if (dom.history.length > 0) singleHistRaw.push([eventName, dom.history]);
      if (dom.ranking.length > 0) singleRankRaw.push([eventName, dom.ranking]);
      collectIds(dom.history, dom.ranking);

      const mem = Math.round(process.memoryUsage.rss() / 1024 / 1024);
      console.log(` ${dom.history.length} records (${((Date.now() - t) / 1000).toFixed(1)}s) [${mem}MB]`);
    }

    // NOTE: Average 基于每轮的 average 值
    for (const [eventId, eventName] of EVENTS_ENTRIES) {
      const t = Date.now();
      process.stdout.write(`  Dominance average: ${eventId}...`);
      let rows = await this.fetchAverageResultsFor(eventId);
      if (rows.length === 0) { console.log(' skip'); continue; }

      const dom = this.computeDominance(rows, 'average');
      rows = null as unknown as RowDataPacket[];
      if (typeof globalThis.gc === 'function') globalThis.gc();

      if (dom.history.length > 0) avgHistRaw.push([eventName, dom.history]);
      if (dom.ranking.length > 0) avgRankRaw.push([eventName, dom.ranking]);
      collectIds(dom.history, dom.ranking);

      const mem = Math.round(process.memoryUsage.rss() / 1024 / 1024);
      console.log(` ${dom.history.length} records (${((Date.now() - t) / 1000).toFixed(1)}s) [${mem}MB]`);
    }

    // NOTE: 所有 event 完成后，一次性查出链接（几十个 person + 几十个 comp）
    const { personMap, compMap } = await this.resolveLinks(allPersonIds, allCompIds);

    const resolveHist = (raw: [string, DomHistoryEntry[]][]): [string, unknown[][]][] =>
      raw.map(([name, entries]) => [name, entries.map(e => [
        e.count, e.improvement, e.days,
        personMap.get(e.personId) ?? e.personId,
        e.firstDate, compMap.get(e.firstCompId) ?? e.firstCompId,
        e.date, compMap.get(e.compId) ?? e.compId,
      ])]);

    const resolveRank = (raw: [string, DomRankingEntry[]][]): [string, unknown[][]][] =>
      raw.map(([name, entries]) => [name, entries.map(e => [
        e.rank, personMap.get(e.personId) ?? e.personId, e.count,
        e.date, compMap.get(e.compId) ?? e.compId,
      ])]);

    return {
      single:  { ranking: resolveRank(singleRankRaw), history: resolveHist(singleHistRaw) },
      average: { ranking: resolveRank(avgRankRaw),     history: resolveHist(avgHistRaw) },
    };
  }

  // NOTE: Single——逐 event 查询 result_attempts（瘦行，无 CONCAT 链接字符串）
  // ⚠️ 不拼 person_link/comp_link——每行省 ~450 bytes，500万行省 ~2GB
  private async fetchSingleAttemptsFor(eventId: string): Promise<RowDataPacket[]> {
    return dbQuery(`
      SELECT
        result.person_id,
        ra.value,
        result.competition_id AS comp_id,
        competition.start_date
      FROM result_attempts ra
      JOIN results result ON result.id = ra.result_id
      JOIN competitions competition ON competition.id = result.competition_id
      WHERE ra.value > 0 AND result.event_id = '${eventId}'
      ORDER BY competition.start_date
    `);
  }

  // NOTE: Average——逐 event 查询每轮 average（瘦行）
  private async fetchAverageResultsFor(eventId: string): Promise<RowDataPacket[]> {
    return dbQuery(`
      SELECT
        result.person_id,
        result.average,
        result.competition_id AS comp_id,
        competition.start_date
      FROM results result
      JOIN competitions competition ON competition.id = result.competition_id
      WHERE result.average > 0 AND result.event_id = '${eventId}'
      ORDER BY competition.start_date
    `);
  }

  // NOTE: 仅为最终结果行（几十行）查询 person/competition 的链接字符串
  private async resolveLinks(personIds: Set<string>, compIds: Set<string>): Promise<{
    personMap: Map<string, string>;
    compMap: Map<string, string>;
  }> {
    const personMap = new Map<string, string>();
    const compMap = new Map<string, string>();

    if (personIds.size > 0) {
      const pids = [...personIds].map(id => `'${id}'`).join(',');
      const rows = await dbQuery(`
        SELECT wca_id,
          CONCAT('[', name, '](https://www.worldcubeassociation.org/persons/', wca_id, ')') AS link
        FROM persons WHERE wca_id IN (${pids}) AND sub_id = 1
      `);
      for (const r of rows) personMap.set(String(r['wca_id']), String(r['link']));
    }

    if (compIds.size > 0) {
      const cids = [...compIds].map(id => `'${id}'`).join(',');
      const rows = await dbQuery(`
        SELECT id,
          CONCAT('[', cell_name, '](https://www.worldcubeassociation.org/competitions/', id, ')') AS link
        FROM competitions WHERE id IN (${cids})
      `);
      for (const r of rows) compMap.set(String(r['id']), String(r['link']));
    }

    return { personMap, compMap };
  }

  // NOTE: 统一计算 dominance 的历史和排名——与 Ruby compute_dominance 1:1 对应
  // 性能优化：per-person 有序数组（二分插入 O(log n)）+ 二分搜索计数
  // NOTE: 瘦行模式——rows 无 person_link/comp_link，返回 ID 以便后续 resolveLinks
  private computeDominance(
    rows: RowDataPacket[], valueCol: string,
  ): { history: DomHistoryEntry[]; ranking: DomRankingEntry[] } {
    const pv = new Map<string, number[]>();
    const pb = new Map<string, number>();
    const pi = new Map<string, { compId: string; date: string }>();

    let maxDom = 0;
    const wrRecords: Array<{
      count: number; personId: string; compId: string; date: string;
    }> = [];
    const firstDom = new Map<string, { compId: string; date: string }>();

    // NOTE: rows 已按 start_date 排序，逐行处理
    // ⚠️ 不建 byDate Map——避免持有全部 rows 引用
    let currentDate = '';

    for (const r of rows) {
      const pid = String(r['person_id']);
      const val = Number(r[valueCol]);
      const dateStr = formatDate(r['start_date']);
      const compId = String(r['comp_id']);

      // 日期切换时检查前一日期组的 dominance
      if (dateStr !== currentDate && currentDate !== '') {
        this.checkDominance(pv, pb, pi, currentDate, maxDom, wrRecords, firstDom);
        if (wrRecords.length > 0 && wrRecords[wrRecords.length - 1].count > maxDom) {
          maxDom = wrRecords[wrRecords.length - 1].count;
        }
      }
      currentDate = dateStr;

      // 二分插入，保持有序
      const arr = pv.get(pid) ?? [];
      if (!pv.has(pid)) pv.set(pid, arr);
      const idx = this.bisectLeft(arr, val);
      arr.splice(idx, 0, val);

      // 更新 best
      const currentBest = pb.get(pid);
      if (!currentBest || val < currentBest) pb.set(pid, val);

      pi.set(pid, { compId, date: dateStr });
    }
    // 最后一个日期组
    if (currentDate !== '') {
      this.checkDominance(pv, pb, pi, currentDate, maxDom, wrRecords, firstDom);
      if (wrRecords.length > 0 && wrRecords[wrRecords.length - 1].count > maxDom) {
        maxDom = wrRecords[wrRecords.length - 1].count;
      }
    }

    // --- 历史：计算 improvement 和 days，然后倒序 ---
    const history: DomHistoryEntry[] = wrRecords.map((r, i) => {
      const improvement = i > 0 ? `+${r.count - wrRecords[i - 1].count}` : '';
      let days: string;
      if (i < wrRecords.length - 1) {
        const nextDate = new Date(wrRecords[i + 1].date);
        const currDate = new Date(r.date);
        days = String(Math.round((nextDate.getTime() - currDate.getTime()) / 86400000));
      } else {
        days = String(Math.round((Date.now() - new Date(r.date).getTime()) / 86400000));
      }
      const first = firstDom.get(r.personId)!;
      return {
        count: r.count, improvement, days,
        personId: r.personId,
        firstCompId: first.compId, firstDate: first.date,
        compId: r.compId, date: r.date,
      };
    }).reverse();

    // --- 排名：复用 pv/pb 最终状态 ---
    const ranking = this.computeRankingFromState(pv, pb, pi);

    return { history, ranking };
  }

  // NOTE: checkDominance——检查当前日期组的 dominance 并记录 WR
  private checkDominance(
    pv: Map<string, number[]>,
    pb: Map<string, number>,
    pi: Map<string, { compId: string; date: string }>,
    date: string,
    maxDom: number,
    wrRecords: Array<{ count: number; personId: string; compId: string; date: string }>,
    firstDom: Map<string, { compId: string; date: string }>,
  ): void {
    let topPid: string | null = null;
    let topBest = Infinity;
    let secondBest = Infinity;

    for (const [pid, best] of pb) {
      if (best < topBest) {
        secondBest = topBest;
        topPid = pid;
        topBest = best;
      } else if (best < secondBest) {
        secondBest = best;
      }
    }

    if (secondBest === Infinity || !topPid) return;

    const arr = pv.get(topPid)!;
    const cnt = this.bisectLeft(arr, secondBest);

    if (cnt > maxDom && cnt > 0) {
      if (!firstDom.has(topPid)) {
        firstDom.set(topPid, { compId: pi.get(topPid)!.compId, date });
      }
      wrRecords.push({
        count: cnt,
        personId: topPid,
        compId: pi.get(topPid)!.compId,
        date,
      });
    }
  }

  // NOTE: 从 pv/pb/pi 最终状态计算当前 dominance Top 10
  private computeRankingFromState(
    pv: Map<string, number[]>,
    pb: Map<string, number>,
    pi: Map<string, { compId: string; date: string }>,
  ): DomRankingEntry[] {
    let topPid: string | null = null;
    let topBest = Infinity;
    let secondBest = Infinity;

    for (const [pid, best] of pb) {
      if (best < topBest) {
        secondBest = topBest;
        topPid = pid;
        topBest = best;
      } else if (best < secondBest) {
        secondBest = best;
      }
    }

    if (secondBest === Infinity) return [];

    const candidates: Array<{ count: number; personId: string; compId: string; date: string }> = [];
    for (const [pid, values] of pv) {
      const othersBest = (pid === topPid) ? secondBest : topBest;
      const cnt = this.bisectLeft(values, othersBest);
      if (cnt <= 0) continue;
      const info = pi.get(pid)!;
      candidates.push({ count: cnt, personId: pid, compId: info.compId, date: info.date });
    }

    candidates.sort((a, b) => b.count - a.count);
    return candidates.slice(0, 10).map((c, i) => ({
      rank: i + 1, personId: c.personId, count: c.count, date: c.date, compId: c.compId,
    }));
  }

  // NOTE: 二分搜索——找到有序数组中第一个 >= target 的位置
  private bisectLeft(arr: number[], target: number): number {
    let lo = 0, hi = arr.length;
    while (lo < hi) {
      const mid = (lo + hi) >>> 1;
      if (arr[mid] < target) lo = mid + 1;
      else hi = mid;
    }
    return lo;
  }
}
